src/spn/algorithms/MEU.py

```diff
@@ -0,0 +1,162 @@
+from spn.structure.Base import get_nodes_by_type, Max, Leaf, Sum, Product, get_topological_order_layers
+
+import numpy as np
+from collections import defaultdict
+import collections
+
+
+def merge_input_vals(l):
+    return np.concatenate(l)
+
+def meu_max(node, parent_result, data=None, lls_per_node=None, rand_gen=None):
+    if len(parent_result) == 0:
+        return None
+
+    parent_result = merge_input_vals(parent_result)
+
+    w_children_log_probs = np.zeros((len(parent_result), len(node.dec_values)))
+    for i, c in enumerate(node.children):
+        w_children_log_probs[:, i] = lls_per_node[parent_result, c.id]
+
+    max_child_branches = np.argmax(w_children_log_probs, axis=1)
+
+    children_row_ids = {}
+
+    for i, c in enumerate(node.children):
+        children_row_ids[c] = parent_result[max_child_branches == i]
+    # print("node and max_child_branches", (node.var_name, max_child_branches))
+    ## decision values at each node
+
+    decision_values = {}
+    decision_values[node.feature_name] = np.column_stack((parent_result, max_child_branches))
+
+    # print("w_children_log_probs",w_children_log_probs)
+    return children_row_ids, decision_values
+
+
+
+node_functions =  get_node_funtions()
+_node_top_down_meu= node_functions[0].copy()
+_node_bottom_up_meu = node_functions[1].copy()
+_node_top_down_meu.update({Max:meu_max})
+_node_bottom_up_meu.update({Sum: sum_likelihood, Product: prod_likelihood, Max:max_likelihood})
+
+
+
+
+def meu(node, input_data, node_top_down_meu=_node_top_down_meu, node_bottom_up_meu=_node_bottom_up_meu, in_place=False):
+    valid, err = is_valid(node)
+    assert valid, err
+    if in_place:
+        data = input_data
+    else:
+        data = np.array(input_data)
+
+    nodes = get_nodes_by_type(node)
+
+    lls_per_node = np.zeros((data.shape[0], len(nodes)))
+
+    # one pass bottom up evaluating the likelihoods
+    # log_likelihood(node, data, dtype=data.dtype, node_log_likelihood=node_bottom_up_meu, lls_matrix=lls_per_node)
+    likelihood(node, data, dtype=data.dtype, node_likelihood=node_bottom_up_meu, lls_matrix=lls_per_node)
+
+    meu_val = lls_per_node[:, 0]
+
+    instance_ids = np.arange(data.shape[0])
+
+    # one pass top down to decide on the max branch until it reaches a leaf; returns  all_result, decisions at each max node for each instance.
+    all_result, all_decisions = eval_spn_top_down_meu(node, node_top_down_meu, parent_result=instance_ids, data=data,
+                                                      lls_per_node=lls_per_node)
+
+    decisions = merge_rows_for_decisions(all_decisions)
+
+    return meu_val, decisions
+
+
+
+def merge_rows_for_decisions(all_decisions=None):
+    """
+    merges different values of same key into one key:value pair
+    :param all_decisions: takes the dictionary of decisions with key as max_node name and values as numpy array with first column as row number of instance, second column as decision.
```

Comment:
since you've mentioned the type of `:param all_decisions:`, may I suggest adding another line which says `:type all_decisions:` which details the type as well.

Reply:
Nice, that breaks it down well. Yep, I'll add that.